presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -354,54 +368,50 @@ private void invokeNoMoreSplitsIfNecessary()
             }
-
-            // TODO: This should use an iterator like the HiveFileIterator
             ListenableFuture<?> lastResult = COMPLETED_FUTURE;
             List<Path> targetPaths = hdfsEnvironment.doAs(
                     hdfsContext.getIdentity().getUser(),
                     () -> getTargetPathsFromSymlink(fs, path));
-            for (Path targetPath : targetPaths) {
-                // The input should be in TextInputFormat.
-                TextInputFormat targetInputFormat = new TextInputFormat();
-                // the splits must be generated using the file system for the target path
-                // get the configuration for the target path -- it may be a different hdfs instance
-                FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(hdfsContext, targetPath);
-                JobConf targetJob = toJobConf(targetFilesystem.getConf());
-                targetJob.setInputFormat(TextInputFormat.class);
-                Optional<Principal> principal = hdfsContext.getIdentity().getPrincipal();
-                if (principal.isPresent()) {
-                    targetJob.set(MRConfig.FRAMEWORK_NAME, MRConfig.CLASSIC_FRAMEWORK_NAME);
-                    targetJob.set(MRConfig.MASTER_USER_NAME, principal.get().getName());
-                }
-                targetInputFormat.configure(targetJob);
-                FileInputFormat.setInputPaths(targetJob, targetPath);
-                InputSplit[] targetSplits = hdfsEnvironment.doAs(
-                        hdfsContext.getIdentity().getUser(),
-                        () -> targetInputFormat.getSplits(targetJob, 0));
-
-                InternalHiveSplitFactory splitFactory = new InternalHiveSplitFactory(
-                        targetFilesystem,
-                        partitionName,
-                        inputFormat,
-                        schema,
-                        partitionKeys,
-                        effectivePredicate,
-                        partitionMatchSupplier,
-                        partition.getTableToPartitionMapping(),
-                        Optional.empty(),
-                        getMaxInitialSplitSize(session),
-                        isForceLocalScheduling(session),
-                        s3SelectPushdownEnabled);
-                lastResult = addSplitsToSource(targetSplits, splitFactory);
-                if (stopped) {
+            ImmutableSet<Path> parents = targetPaths.stream()
+                    .map(Path::getParent)
+                    .distinct()
+                    .collect(toImmutableSet());
+            InputFormat<?, ?> fileInputFormat = mapSymlinkSerdeToInputFormat(toJobConf(configuration), getDeserializerClassName(schema));
+            if (parents.size() == 1 && !recursiveDirWalkerEnabled &&
+                    !targetPaths.stream().anyMatch((targetPath) -> targetPath.getName().endsWith("/"))) {
```

Comment:
I was thinking it could be a minor optimization to avoid `buildManifestFileIterator` if one of the paths "looks" like a directory, but we can always do the explicit check.